File: packages/android/sdk/src/main/kotlin/com/clienttools/sdk/inspector/InspectorApiHandler.kt
package com.clienttools.sdk.inspector

import android.util.Base64
import android.util.Log
import fi.iki.elonen.NanoHTTPD
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

class InspectorApiHandler(
    private val fileStore: InspectorFileStore,
    private val imageFileStore: ImageFileStore,
    private val getTopViewModel: () -> InspectorViewModel?
) {
    private val TAG = "InspectorApiHandler"
    private val json = Json { ignoreUnknownKeys = true }

    // ── WebView HTML ──────────────────────────────────────────────────────────

    fun handlePushHtml(body: String): NanoHTTPD.Response = try {
        val obj = json.parseToJsonElement(body).jsonObject
        val tag = obj["tag"]?.jsonPrimitive?.content ?: return error(400, "Missing tag")
        val html = obj["html"]?.jsonPrimitive?.content ?: return error(400, "Missing html")
        val timestamp = obj["timestamp"]?.jsonPrimitive?.content ?: fileStore.generateTimestamp()

        val saved = fileStore.saveHtmlFile(tag, timestamp, html)
            ?: return error(500, "Failed to save file")

        getTopViewModel()?.let { vm ->
            vm.webView.value = vm.webView.value.copy(
                currentFile = FileInfo(tag, timestamp, saved.fileUrl),
                isVisible = true
            )
        }

        ok("""{"code":0,"message":"success","data":{"tag":"$tag","timestamp":"$timestamp","filePath":"${saved.fileUrl}","fileSize":${html.length}}}""")
    } catch (e: Exception) {
        Log.e(TAG, "pushHtml error", e)
        error(400, "Invalid request: ${e.message}")
    }

    fun handleShow(body: String): NanoHTTPD.Response = try {
        val obj = json.parseToJsonElement(body).jsonObject
        val tag = obj["tag"]?.jsonPrimitive?.content ?: return error(400, "Missing tag")
        val timestamp = obj["timestamp"]?.jsonPrimitive?.content ?: return error(400, "Missing timestamp")

        val fileUrl = fileStore.getFilePath(tag, timestamp) ?: return error(404, "File not found")

        getTopViewModel()?.let { vm ->
            vm.webView.value = vm.webView.value.copy(
                currentFile = FileInfo(tag, timestamp, fileUrl),
                isVisible = true
            )
            val s = vm.webView.value
            return ok("""{"code":0,"message":"success","data":{"tag":"$tag","timestamp":"$timestamp","opacity":${s.opacity},"offsetX":${s.offsetX},"offsetY":${s.offsetY}}}""")
        }
        ok("""{"code":0,"message":"success","data":{"tag":"$tag","timestamp":"$timestamp"}}""")
    } catch (e: Exception) {
        Log.e(TAG, "show error", e)
        error(500, "Internal error: ${e.message}")
    }

    fun handleGetFiles(currentFile: FileInfo?): NanoHTTPD.Response = try {
        val vmCurrentFile = getTopViewModel()?.webView?.value?.currentFile ?: currentFile
        val files = fileStore.getAllFiles()
        val filesJson = files.joinToString(",") { f ->
            val isCurrent = vmCurrentFile?.tag == f.tag && vmCurrentFile?.timestamp == f.timestamp
            val size = java.io.File(f.fileUrl.removePrefix("file://")).length()
            """{"tag":"${f.tag}","timestamp":"${f.timestamp}","size":$size,"isCurrent":$isCurrent}"""
        }
        ok("""{"code":0,"data":{"files":[$filesJson]}}""")
    } catch (e: Exception) {
        Log.e(TAG, "getFiles error", e)
        error(500, "Internal error: ${e.message}")
    }

    // ── Image ─────────────────────────────────────────────────────────────────

    fun handlePushImage(body: String): NanoHTTPD.Response = try {
        val obj = json.parseToJsonElement(body).jsonObject
        val tag = obj["tag"]?.jsonPrimitive?.content ?: return error(400, "Missing tag")
        val imageBase64 = obj["image"]?.jsonPrimitive?.content ?: return error(400, "Missing image")
        val ext = obj["ext"]?.jsonPrimitive?.content?.lowercase() ?: "png"
        val timestamp = obj["timestamp"]?.jsonPrimitive?.content ?: imageFileStore.generateTimestamp()

        val bytes = try {
            Base64.decode(imageBase64, Base64.DEFAULT)
        } catch (e: Exception) {
            return error(400, "Invalid base64: ${e.message}")
        }

        val saved = imageFileStore.saveImage(tag, timestamp, bytes, ext)
            ?: return error(500, "Failed to save image")

        getTopViewModel()?.let { vm ->
            vm.image.value = vm.image.value.copy(currentImage = saved, isVisible = true)
        }

        ok("""{"code":0,"message":"success","data":{"tag":"$tag","timestamp":"$timestamp","filePath":"${saved.filePath}","fileSize":${bytes.size}}}""")
    } catch (e: Exception) {
        Log.e(TAG, "pushImage error", e)
        error(400, "Invalid request: ${e.message}")
    }

    fun handleShowImage(body: String): NanoHTTPD.Response = try {
        val obj = json.parseToJsonElement(body).jsonObject
        val tag = obj["tag"]?.jsonPrimitive?.content ?: return error(400, "Missing tag")
        val timestamp = obj["timestamp"]?.jsonPrimitive?.content ?: return error(400, "Missing timestamp")

        val filePath = imageFileStore.getFilePath(tag, timestamp) ?: return error(404, "Image not found")
        val ext = java.io.File(filePath).extension.lowercase()

        getTopViewModel()?.let { vm ->
            vm.image.value = vm.image.value.copy(
                currentImage = ImageInfo(tag, timestamp, filePath, ext),
                isVisible = true
            )
            val s = vm.image.value
            return ok("""{"code":0,"message":"success","data":{"tag":"$tag","timestamp":"$timestamp","opacity":${s.opacity},"offsetX":${s.offsetX},"offsetY":${s.offsetY}}}""")
        }
        ok("""{"code":0,"message":"success","data":{"tag":"$tag","timestamp":"$timestamp"}}""")
    } catch (e: Exception) {
        Log.e(TAG, "showImage error", e)
        error(500, "Internal error: ${e.message}")
    }

    fun handleGetImages(currentImage: ImageInfo?): NanoHTTPD.Response = try {
        val vmCurrentImage = getTopViewModel()?.image?.value?.currentImage ?: currentImage
        val images = imageFileStore.getAllImages()
        val imagesJson = images.joinToString(",") { img ->
            val isCurrent = vmCurrentImage?.tag == img.tag && vmCurrentImage?.timestamp == img.timestamp
            val size = java.io.File(img.filePath).length()
            """{"tag":"${img.tag}","timestamp":"${img.timestamp}","ext":"${img.ext}","size":$size,"isCurrent":$isCurrent}"""
        }
        ok("""{"code":0,"data":{"images":[$imagesJson]}}""")
    } catch (e: Exception) {
        Log.e(TAG, "getImages error", e)
        error(500, "Internal error: ${e.message}")
    }

    // ── 共用：hide / adjust ────────────────────────────────────────────────────

    fun handleHide(body: String = "{}"): NanoHTTPD.Response = try {
        val obj = runCatching { json.parseToJsonElement(body).jsonObject }.getOrNull()
        val typeStr = obj?.get("type")?.jsonPrimitive?.content
        val vm = getTopViewModel()
        when {
            typeStr == "image" -> vm?.image?.value = vm?.image?.value?.copy(isVisible = false) ?: ImageState()
            typeStr == "webview" -> vm?.webView?.value = vm?.webView?.value?.copy(isVisible = false) ?: WebViewState()
            else -> when (vm?.activeTab?.value) {
                ActiveTab.IMAGE -> vm.image.value = vm.image.value.copy(isVisible = false)
                else -> vm?.webView?.value = vm?.webView?.value?.copy(isVisible = false) ?: WebViewState()
            }
        }
        ok("""{"code":0,"message":"success"}""")
    } catch (e: Exception) {
        Log.e(TAG, "hide error", e)
        error(500, "Internal error: ${e.message}")
    }

    fun handleAdjust(body: String): NanoHTTPD.Response = try {
        val obj = json.parseToJsonElement(body).jsonObject
        val typeStr = obj["type"]?.jsonPrimitive?.content
        val dx = obj["offsetX"]?.jsonPrimitive?.content?.toIntOrNull() ?: 0
        val dy = obj["offsetY"]?.jsonPrimitive?.content?.toIntOrNull() ?: 0
        val opacity = obj["opacity"]?.jsonPrimitive?.content?.toFloatOrNull()

        val vm = getTopViewModel()
        val isImage = typeStr == "image" || (typeStr == null && vm?.activeTab?.value == ActiveTab.IMAGE)

        if (isImage) {
            val s = vm?.image?.value ?: ImageState()
            val newState = s.copy(
                offsetX = s.offsetX + dx,
                offsetY = s.offsetY + dy,
                opacity = opacity?.coerceIn(0f, 1f) ?: s.opacity
            )
            vm?.image?.value = newState
            ok("""{"code":0,"data":{"offsetX":${newState.offsetX},"offsetY":${newState.offsetY},"opacity":${newState.opacity}}}""")
        } else {
            val s = vm?.webView?.value ?: WebViewState()
            val newState = s.copy(
                offsetX = s.offsetX + dx,
                offsetY = s.offsetY + dy,
                opacity = opacity?.coerceIn(0f, 1f) ?: s.opacity
            )
            vm?.webView?.value = newState
            ok("""{"code":0,"data":{"offsetX":${newState.offsetX},"offsetY":${newState.offsetY},"opacity":${newState.opacity}}}""")
        }
    } catch (e: Exception) {
        Log.e(TAG, "adjust error", e)
        error(500, "Internal error: ${e.message}")
    }

    // ── DOM 查询 ────────────────────────────────────────────────────────────────

    private val domQueryService = DomQueryService(timeoutMs = 3000L)

    suspend fun handleDomAll(webView: android.webkit.WebView?): NanoHTTPD.Response {
        if (webView == null) return domError(3, "webview not ready")
        val vm = getTopViewModel()
        val offsetX = vm?.webView?.value?.offsetX ?: 0
        val offsetY = vm?.webView?.value?.offsetY ?: 0
        return try {
            val nodes = domQueryService.queryAll(webView, offsetX, offsetY)
            val nodesJson = nodes.joinToString(",") { n ->
                """{"id":"${n.id}","tagName":"${n.tagName}","x":${n.x},"y":${n.y},"width":${n.width},"height":${n.height},"text":${org.json.JSONObject.quote(n.text)}}"""
            }
            ok("""{"code":0,"data":{"count":${nodes.size},"nodes":[$nodesJson]}}""")
        } catch (e: kotlinx.coroutines.TimeoutCancellationException) {
            domError(2, "timeout")
        } catch (e: Exception) {
            Log.e(TAG, "domAll error", e)
            domError(1, "parse error")
        }
    }

    suspend fun handleDomById(webView: android.webkit.WebView?, id: String): NanoHTTPD.Response {
        if (webView == null) return domError(3, "webview not ready")
        val vm = getTopViewModel()
        val offsetX = vm?.webView?.value?.offsetX ?: 0
        val offsetY = vm?.webView?.value?.offsetY ?: 0
        return try {
            val node = domQueryService.queryById(webView, id, offsetX, offsetY)
                ?: return domError(1, "not found")
            ok("""{"code":0,"data":{"id":"${node.id}","tagName":"${node.tagName}","x":${node.x},"y":${node.y},"width":${node.width},"height":${node.height},"text":${org.json.JSONObject.quote(node.text)}}}""")
        } catch (e: kotlinx.coroutines.TimeoutCancellationException) {
            domError(2, "timeout")
        } catch (e: Exception) {
            Log.e(TAG, "domById error", e)
            domError(1, "parse error")
        }
    }

    private fun domError(code: Int, message: String) = ok("""{"code":$code,"message":"$message"}""")

    // ── 内部工具 ───────────────────────────────────────────────────────────────

    private fun ok(json: String) = NanoHTTPD.newFixedLengthResponse(
        NanoHTTPD.Response.Status.OK, "application/json", json
    )

    private fun error(code: Int, message: String): NanoHTTPD.Response {
        val status = when {
            code == 404 -> NanoHTTPD.Response.Status.NOT_FOUND
            code >= 500 -> NanoHTTPD.Response.Status.INTERNAL_ERROR
            else        -> NanoHTTPD.Response.Status.BAD_REQUEST
        }
        return NanoHTTPD.newFixedLengthResponse(status, "application/json",
            """{"code":$code,"message":"$message"}""")
    }
}

File: clients/ios/sdk/Sources/HttpServer/Inspector/InspectorApiHandler.swift
import Foundation
import Network
import SwiftProtobuf

class InspectorApiHandler {

    private let viewModel: InspectorViewModel
    private let imageFileStore: ImageFileStore
    private let sendProtoFn: (SwiftProtobuf.Message, Int, NWConnection) -> Void
    private let sendErrorFn: (Int32, String, Int, NWConnection) -> Void

    init(
        viewModel: InspectorViewModel,
        imageFileStore: ImageFileStore,
        sendProto: @escaping (SwiftProtobuf.Message, Int, NWConnection) -> Void,
        sendError: @escaping (Int32, String, Int, NWConnection) -> Void
    ) {
        self.viewModel = viewModel
        self.imageFileStore = imageFileStore
        self.sendProtoFn = sendProto
        self.sendErrorFn = sendError
    }

    // MARK: - push-image

    func handlePushImage(_ body: Data, connection: NWConnection) {
        guard let req = try? Clienttools_PushImageRequest(serializedBytes: body) else {
            sendErrorFn(400, "Invalid request", 400, connection); return
        }
        let bytes = Data(req.image)
        let timestamp = req.timestamp.isEmpty ? imageFileStore.generateTimestamp() : req.timestamp
        let ext = req.ext.isEmpty ? "png" : req.ext
        guard let saved = imageFileStore.saveImage(tag: req.tag, timestamp: timestamp, bytes: bytes, ext: ext) else {
            sendErrorFn(500, "Failed to save image", 500, connection); return
        }
        viewModel.imageState = ImageState(
            currentImage: saved, isVisible: true,
            offsetX: viewModel.imageState.offsetX,
            offsetY: viewModel.imageState.offsetY,
            opacity: viewModel.imageState.opacity
        )
        viewModel.activeTab = .image
        var result = Clienttools_PushImageResult()
        result.tag = req.tag; result.timestamp = timestamp
        result.filePath = saved.filePath; result.fileSize = Int64(bytes.count)
        var resp = Clienttools_PushImageResponse()
        resp.meta = okMeta(); resp.data = result
        sendProtoFn(resp, 200, connection)
    }

    // MARK: - show-image

    func handleShowImage(_ body: Data, connection: NWConnection) {
        guard let req = try? Clienttools_ShowImageRequest(serializedBytes: body) else {
            sendErrorFn(400, "Invalid request", 400, connection); return
        }
        guard let filePath = imageFileStore.getFilePath(tag: req.tag, timestamp: req.timestamp) else {
            sendErrorFn(404, "Image not found", 404, connection); return
        }
        let ext = URL(fileURLWithPath: filePath).pathExtension.lowercased()
        let info = ImageInfo(tag: req.tag, timestamp: req.timestamp, filePath: filePath, ext: ext)
        viewModel.imageState = ImageState(
            currentImage: info, isVisible: true,
            offsetX: viewModel.imageState.offsetX,
            offsetY: viewModel.imageState.offsetY,
            opacity: viewModel.imageState.opacity
        )
        viewModel.activeTab = .image
        let s = viewModel.imageState
        var result = Clienttools_ShowImageResult()
        result.tag = req.tag; result.timestamp = req.timestamp
        result.opacity = s.opacity; result.offsetX = s.offsetX; result.offsetY = s.offsetY
        var resp = Clienttools_ShowImageResponse()
        resp.meta = okMeta(); resp.data = result
        sendProtoFn(resp, 200, connection)
    }

    // MARK: - images

    func handleGetImages(connection: NWConnection) {
        let currentImage = viewModel.imageState.currentImage
        let images = imageFileStore.getAllImages()
        let items: [Clienttools_ImageItem] = images.map { img in
            let isCurrent = img.tag == currentImage?.tag && img.timestamp == currentImage?.timestamp
            let size = (try? FileManager.default.attributesOfItem(atPath: img.filePath)[.size] as? Int) ?? 0
            var item = Clienttools_ImageItem()
            item.tag = img.tag; item.timestamp = img.timestamp
            item.ext = img.ext; item.size = Int64(size); item.isCurrent = isCurrent
            return item
        }
        var result = Clienttools_ImageListResult()
        result.images = items
        var resp = Clienttools_ImageListResponse()
        resp.meta = okMeta(); resp.data = result
        sendProtoFn(resp, 200, connection)
    }

    // MARK: - hide

    func handleHide(_ body: Data, connection: NWConnection) {
        let req = (try? Clienttools_HideRequest(serializedBytes: body)) ?? Clienttools_HideRequest()
        switch req.type {
        case "image":
            viewModel.imageState.isVisible = false
        case "webview":
            viewModel.webViewState.isVisible = false
        default:
            if viewModel.activeTab == .image {
                viewModel.imageState.isVisible = false
            } else {
                viewModel.webViewState.isVisible = false
            }
        }
        var resp = Clienttools_SimpleResponse()
        resp.meta = okMeta()
        sendProtoFn(resp, 200, connection)
    }

    // MARK: - adjust

    func handleAdjust(_ body: Data, connection: NWConnection) {
        guard let req = try? Clienttools_InspectorAdjustRequest(serializedBytes: body) else {
            sendErrorFn(400, "Invalid request", 400, connection); return
        }
        let isImage = req.type == "image" || (req.type.isEmpty && viewModel.activeTab == .image)
        var result = Clienttools_InspectorAdjustResult()
        if isImage {
            var s = viewModel.imageState
            s.offsetX += req.offsetX; s.offsetY += req.offsetY
            if req.opacity > 0 { s.opacity = min(max(req.opacity, 0), 1) }
            viewModel.imageState = s
            result.offsetX = s.offsetX; result.offsetY = s.offsetY; result.opacity = s.opacity
        } else {
            var s = viewModel.webViewState
            s.offsetX += req.offsetX; s.offsetY += req.offsetY
            if req.opacity > 0 { s.opacity = min(max(req.opacity, 0), 1) }
            viewModel.webViewState = s
            result.offsetX = s.offsetX; result.offsetY = s.offsetY; result.opacity = s.opacity
        }
        var resp = Clienttools_InspectorAdjustResponse()
        resp.meta = okMeta(); resp.data = result
        sendProtoFn(resp, 200, connection)
    }

    // MARK: - private

    private func okMeta() -> Clienttools_ResponseMeta {
        var meta = Clienttools_ResponseMeta()
        meta.code = 0; meta.message = "success"
        return meta
    }
}
